### ISSUE TITLE
[$1000] Attachments- App crashes when user navigate attachments

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#24337]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Launch app
2. Tap on any report
3. Tap plus button to add few image/attachments 
4. Tap on a image
5. Use side arrow to navigate and view attachments

## Expected Result:
When user navigate attachments using side arrows, he must be able to view all attachments without problem

## Actual Result:
When user navigate attachments using side arrows, the app crashes while trying to view below attached images

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.52.1

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/5a4629f3-2ce7-40c6-838c-c07dea33c41e

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01938fb39d08874250</li>
        <li>Upwork Job ID: 1712662082191142912</li>
        <li>Last Price Increase: 2023-11-27</li>
    </ul>
</details>

**Comment by puneetlath** [comment]
@lanitochka17 is this an android-only issue? I'm not able to reproduce it on iOS.

**Comment by kosmydel** [comment]
> > Hey, it might be related to [this issue](https://github.com/Expensify/App/issues/21133).
> > Your workspace name contains HTML characters (`&`), which is also the cause of the mentioned issue. [More details in my proposal.](https://github.com/Expensify/App/issues/21133#issuecomment-1666457954)
> > Can you try to reproduce it on a new workspace, with a name without any special characters?
> 
> If so, it should be reproducible on production as well and not deploy blocker.

You are right. It was just a guess.

**Comment by Li357** [comment]
Can't reproduce on iOS and Android either. I'm going to close this out for now, if you can provide extra details that would be great!

**Comment by lanitochka17** [comment]
@Li357 Issue is still reproducible on Android
I have added an attachment that is causing the app to crash and there is one attachment that needs to be added that is larger than 10MB

https://github.com/Expensify/App/assets/78819774/21137210-9e45-48f5-8af5-df9f94fb9a17

![Bug6159108_Profile_image](https://github.com/Expensify/App/assets/78819774/f49d8f55-71b8-48d3-9845-e580fc888380)

**Comment by GabiHExpensify** [comment]
This issue hasn't been commented on in two weeks, should it still be labeled an `hourly`? If it's something that can go two weeks without progress without many consequences, then `hourly` seems like overkill. And if it's something that needs to be fixed within the next hour, then could you comment on progress @Li357? thanks!

**Comment by GabiHExpensify** [comment]
Sounds like this is no longer a deploy blocker, so I'm updating it to a `daily`.

**Comment by Li357** [comment]
Going to look into this tomorrow. I hadn't gotten to this because I had some wave 8 things to prioritize and hadn't been able to build Android for the past few days. Will reprioritize!

**Comment by Li357** [comment]
Fighting android emulator issues, but testing this now.

**Comment by akinwale** [comment]
I spent some time trying to figure this out and was finally able to reproduce (you have to test with the attached image in https://github.com/Expensify/App/issues/24337#issuecomment-1729458221). The crash is due to an out of memory error due to the number of pixels in the image, I think it's some insanely huge number (9000x9000, iirc but it may be higher). The solution would probably be to try to scale the image down before displaying it.

**Comment by akinwale** [comment]
> @akinwale I suspected something like that, but the attached image is only 814KB for me? We have an attachment maximum set to 24MB, maybe we should consider lowering that? Since the point of the modal is to show the full-res image after clicking on a lower-res thumbnail.

It's not the file size, but the amount of memory required to display the number of pixels in the image. 

So in this case, using 4 bytes per pixel, displaying 9000x9000 pixels would require 4x9000x9000 which equates to 324MB. The Java heap size is probably lower than that, hence the crash (I'm guessing here).

The image scaling would have to be to reduce the total number of pixels, not the actual file size of fhe image.

**Comment by Li357** [comment]
> So in this case, using 4 bytes per pixel (assuming ARGB), displaying 9000x9000 pixels would require 4x9000x9000 which equates to 324MB.

Great point! I guess the filesize is small since the image is mostly solid colors. Would you be interested in investigating the solution further? If not, I'll take this over.

**Comment by akinwale** [comment]
I don't plan on submitting a proposal for this one as I haven't figured out a viable solution yet.

**Comment by Li357** [comment]
Alright, since I've got limited bandwidth, I'm going to take this external and see if we can get proposals here

**Comment by Li357** [comment]
Not overdue, waiting for proposals. Will up the price if we don't get enough interest here

**Comment by Li357** [comment]
No proposals yet. Moving to weekly to increase bounty.

**Comment by paultsimura** [comment]
It might be worth waiting on https://github.com/Expensify/App/pull/28640 and https://github.com/Expensify/App/issues/27237. 
The attachments behavior is quite unstable now, which should be improved after those issues are resolved. Also, an issue similar to this (https://github.com/Expensify/App/issues/29109) was put on hold for https://github.com/Expensify/App/issues/27237, which currently has proposals being reviewed.

**Comment by paultsimura** [comment]
@Li357 please check this comment: https://github.com/Expensify/App/issues/24337#issuecomment-1773715499

**Comment by Li357** [comment]
@paultsimura Do you think this issue will get absorbed by those? This instability would require some extra code to show thumbnails instead of the full image

**Comment by paultsimura** [comment]
@Li357 my bad, I wasn't attentive enough to the discussion above. There just was a storm of "Attachments are crashing" dupe issues at the time I posted my initial comment. For this specific scenario, I don't think the linked issues will help.

**Comment by Li357** [comment]
Bounty increased, I think this is valuable to work on because images get to this size pretty easily

**Comment by chrispader** [comment]
> I spent some time trying to figure this out and was finally able to reproduce (you have to test with the attached image in [#24337 (comment)](https://github.com/Expensify/App/issues/24337#issuecomment-1729458221)). The crash is due to an out of memory error due to the number of pixels in the image, I think it's some insanely huge number (9000x9000, iirc but it may be higher). The solution would probably be to try to scale the image down before displaying it.

@ArekChr was working on a [PR in `react-native-fast-image`](https://github.com/Expensify/react-native-fast-image/pull/6) a few months ago, which should fix this problem by downsampling images.

@ArekChr do you fix your PR could fix this problem and what's the status on the PR? :)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by giltron** [comment]
This was proposed as solution with [31204](https://github.com/Expensify/App/issues/31204) that can fix crash with reproducible image attached

## Proposal

### Please re-state the problem that we are trying to solve in this issue.

App crashes after attempting to view a large uploaded image attachment.

### What is the root cause of that problem?

The image size, which is 7724 x 5148, is too large for the rendering on the screen.
This causes the app to crash with the following in logcat on Android:

```
 Canvas: trying to draw too large(159052608bytes) bitmap.
```

This image is too large for the container it is being rendered on the device.


### What changes do you think we should make in order to solve the problem?

Make sure the image height and size are not out of bounds (for the canvas size) what is supported by the device display.

Within [AttachmentCarouselPage](https://github.com/Expensify/App/blob/5e7022d8cc519ab8787aa0e0d70d5692c6c27945/src/components/Attachments/AttachmentCarousel/Pager/AttachmentCarouselPage.js), the <Image> components' onLoad should size `dimensions` similar to existing approach in [ImageView](https://github.com/Expensify/App/blob/5e7022d8cc519ab8787aa0e0d70d5692c6c27945/src/components/ImageView/index.native.js#L42).  This component is used for the initial image preview during adding the attachment:

```
  // Resize the image to max dimensions possible on the Native platforms to prevent crashes on Android. To keep the same behavior, apply to IOS as well.
        const maxDimensionsScale = 11;
        imageZoomWidth = Math.min(imageZoomWidth, roundedContainerWidth * maxDimensionsScale);
        imageZoomHeight = Math.min(imageZoomHeight, roundedContainerHeight * maxDimensionsScale);
```


The solution is to adjust the imageHeight and imageWidth calculation within the onLoad events in [AttachmentCarouselPage](https://github.com/Expensify/App/blob/5e7022d8cc519ab8787aa0e0d70d5692c6c27945/src/components/Attachments/AttachmentCarousel/Pager/AttachmentCarouselPage.js( with similar approach as [ImageView](https://github.com/Expensify/App/blob/5e7022d8cc519ab8787aa0e0d70d5692c6c27945/src/components/ImageView/index.native.js#L42). [AttachmentCarouselPage](https://github.com/Expensify/App/blob/5e7022d8cc519ab8787aa0e0d70d5692c6c27945/src/components/Attachments/AttachmentCarousel/Pager/AttachmentCarouselPage.js) to ensure the maximum dimensions are not too large for the canvas:


```
imageWidth = Math.min(imageWidth, canvasWidth * maxDimensionsScale);
imageHeight = Math.min(imageHeight, canvasHeight * maxDimensionsScale);
```

### What alternative solutions did you explore? (Optional)

Additionally, if there are an intended  maximum dimensions for general image attachments this validation can be added before the image is uploaded.  (personally don't like this solution unless there is a different need to scale down images)


In `AttachmentPicker`, we can add a function to check the size dimensions for an image (can check by height, width, and overall size):
```
isValidResolution(image) {
        return getImageResolution(image).then(
            (resolution) =>
                resolution.height >= CONST.ATTACHMENT_MIN_HEIGHT_PX &&
                resolution.width >= CONST.ATTACHMENT_WIDTH_PX &&
                resolution.height <= CONST.ATTACHMENT_MAX_HEIGHT_PX &&
                resolution.width <= CONST.ATTACHMENT_MAX_WIDTH_PX,
        );
    }

```

For native Android/iOS, we can also configure react-native-image-picker to automatically resize the image:
```
maxWidth
maxHeight
```

For informing the user, a similar message could be displayed (if it is not desirable to automatically resize) as with the Avatar image upload


![282525660-d1d19d6d-e6dc-4bad-8829-e6bfafa3d0d5](https://github.com/Expensify/App/assets/2394819/a2d794b5-4954-4edc-8ff6-a1e4a852a19e)

**Comment by chrispader** [comment]
@giltron There should be no need for this change anymore. We're completely overhauling the native `ImageView` component with a new `MultiGestureCanvas`/`Lightbox` component in [this PR](https://github.com/Expensify/App/pull/31308).

The crash with large images has also been fixed by a patch: https://github.com/Expensify/App/issues/28159#issuecomment-1821116608

---

cc @akinwale @Li357 @mvtglobally

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)
